File: Graph/prims.cpp
#include<iostream>
#include<limits.h>

using namespace std;


// Time Complexity is E*(logV);

/*
Check kruskal.cpp for spanning tree definitions.

Prim's algorithm is a minimum spanning tree algorithm that takes a graph as input
and finds the subset of the edges of that graph which form a tree that includes every vertex and
has the minimum sum of weights among all the trees that can be formed from the graph

Approach:
Take any vertex as starting vertex.
At each vertex, all the edges are considered and the one which has minimum weight and doesn't form a cycle is choosen.
*/

void print_graph(int** graph, int v);
int** prims(int** graph, int v);
bool detect_cycle(int** graph, int v, int index, bool *visited, int **visited_edges);


int main() {
	int n=0,e;
	cout << "Enter the number of vertices: ";
	cin >> n;
	int** graph;
	graph=new int*[n];
	for(int i=0;i<n;i++) {
		graph[i]=new int[n];
	}
	for(int i=0;i<n;i++) {
		for(int j=0;j<n;j++) {
			graph[i][j]=0;
		}
	}
	cout << "Enter the edges: " <<endl;
	int x;
	int k=0;
	//Graph Initialisation
	while(k!=e) {
		cout << "\n\n******* MENU *******" <<endl;
		cout <<"Press \'1\' to enter edge." <<endl;
		cout <<"Press \'2\' to print the current graph." <<endl;
		cout <<"Press \'3\' to exit entering edge." <<endl;
		cin >> x;
		if(x==3) {
			break;
		}
		else if(x==2) {
			print_graph(graph,n);
		}
		else if(x==1) {
			int u,v,w;
			cout << "Enter the first vertex of the edge: ";
			cin >> u;
			cout << "Enter the second vertex of the edge: ";
			cin >>v;
			cout << "Enter the weight of the edge: ";
			cin >> w;
			if(u<0||v<0||u>=n||v>=n) {
				cout << "Invalid Input. Try Again!" << endl;
				continue;
			}
			graph[u][v]=w;
			graph[v][u]=w;
			k++;
		}
		else {
			cout << "Invalid Input. Try Again!" <<endl;
			continue;
		}
	}
	print_graph(graph,n);
	int** result_tree=prims(graph,n);
	cout << "\n\nResulting Spanning Tree: " << endl;
	print_graph(result_tree,n);
}

void print_graph(int** graph, int v) {
	cout <<"\n\nCurrent Graph: " <<endl;
	for(int i=0;i<v;i++) {
		for(int j=0;j<v;j++) {
			cout << graph[i][j] << "\t";
		}
		cout << endl;
	}
}


int** prims(int** graph, int v) {
	//Initialised MST
	int **tree=new int*[v];
	for(int i=0;i<v;i++) {
		tree[i]=new int[v];
	}
	for(int i=0;i<v;i++) {
		for(int j=0;j<v;j++) {
			tree[i][j]=0;
		}
	}
	//For each vertex
	for(int i=0;i<v;i++) {
		//Initialised MIN INDEX & MIN.
		int min_index=0;
		int min=INT_MAX;
		//For each edge from vertex i.
		for(int j=0;j<v;j++) {
			if(graph[i][j]!=0) {
				//Edge is included in MST.
				tree[i][j]=graph[i][j];
				tree[j][i]=graph[i][j];
				//Initialised visited vertex array && visited edges array.
				bool *visited=new bool[v];
				int** visited_edges=new int*[v];
				for(int x=0;x<v;x++) {
					visited[x]=false;
					visited_edges[x]=new int[v];
				}
				for(int x=0;x<v;x++) {
					for(int y=0;y<v;y++) {
						visited_edges[x][y]=0;
					}
				}
				//Detect Cycle
				if(!detect_cycle(tree,v,i,visited,visited_edges)) {
					//If there is no cycle, compare with min-weight edge.
					if(min>graph[i][j]) {
						min=graph[i][j];
						min_index=j; 
					}
				}
				//Edge is discarded since minimum weight edge has to been included.
				tree[i][j]=0;
				tree[j][i]=0;
			}
		}
		//Minimum weight edge included.
		tree[i][min_index]=min;
		tree[min_index][i]=min;
	}
	return tree;
}

//Modified DFS
bool detect_cycle(int** graph, int v, int index, bool *visited, int **visited_edges) {
	//mark vertex as visited.
	visited[index]=true;
	//DFS
	for(int i=0;i<v;i++) {
		if(graph[index][i]!=0&&!visited[i]) {
			visited_edges[index][i]=1;
			visited_edges[i][index]=1;
			//If cycle is detected, return true immediately.
			if(detect_cycle(graph,v,i,visited,visited_edges)) {
				return true; 
			}
		}
	}
	//check whetether edge is not in visited edges array but the index between the edges are already visited or not.
	for(int i=0;i<v;i++) {
		if(graph[index][i]!=0&&visited[i]&&visited_edges[i][index]!=1&&visited_edges[index][i]!=1) {
			cout << "\nCycle Detected: "<< index << "\t" << i << endl;
			//If yes, return true else return false.
			return true;
		}
	}
	return false;
}

File: Graph/Prims.java
/*
 * Prim's Algorithm for finding the Minimum Spanning Tree (MST) of a connected, undirected, weighted graph.
 *
 * Explanation:
 * - The graph is represented using an adjacency list.
 * - The algorithm starts from vertex 1 (0-th index in array).
 * - It uses a priority queue (min-heap) to always pick the smallest edge that connects a visited node to an unvisited one.
 * - The process continues until all vertices are included in the MST or until no more valid edges are available.
 *
 * Time Complexity:
 * - Let V be the number of vertices and E the number of edges.
 * - Each vertex is added to the MST once, and each edge is processed at most once.
 * - Adding/removing from the priority queue takes O(log E) time.
 * - So overall complexity is: O(E log E) or O(E log V), depending on implementation and edge distribution.
 *
 * Space Complexity:
 * - Adjacency list: O(V + E)
 * - Priority queue: O(E)
 * - Visited array and MST list: O(V)
 * - Total: O(V + E)
 */

import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Prims {
  private static final Scanner sc = new Scanner(System.in);

  public static void main(String[] args) {
    System.out.println("Enter the number of vertices: ");
    int v = sc.nextInt();
    System.out.println("Enter the number of edges: ");
    int e = sc.nextInt();

    ArrayList<GraphEdge>[] adjList = new ArrayList[v];
    for (int i = 0; i < v; i++) {
      adjList[i] = new ArrayList<>();
    }

    for (int i = 0; i < e; i++) {
      System.out.println("Enter details of the edge " + (i + 1));
      System.out.println("Enter the first vertex of the edge: ");
      int v1 = sc.nextInt();
      System.out.println("Enter the second vertex of the edge: ");
      int v2 = sc.nextInt();
      System.out.println("Enter the weight of the edge: ");
      int w = sc.nextInt();

      if (v1 < 1 || v1 > v || v2 < 1 || v2 > v) {
        System.out.println("Invalid vertex number. Try again.");
        i--;
        continue;
      }

      adjList[v1 - 1].add(new GraphEdge(v1, v2, w));
      adjList[v2 - 1].add(new GraphEdge(v2, v1, w));
    }

    boolean[] visited = new boolean[v];
    PriorityQueue<GraphEdge> queue = new PriorityQueue<>();

    visited[0] = true;
    for (GraphEdge edge : adjList[0]) {
      queue.offer(edge);
    }

    ArrayList<GraphEdge> mst = new ArrayList<>();

    while (!queue.isEmpty() && mst.size() != v - 1) {
      GraphEdge edge = queue.poll();
      if (!visited[edge.v2 - 1]) {
        mst.add(edge);
        visited[edge.v2 - 1] = true;
        for (GraphEdge nextEdge : adjList[edge.v2 - 1]) {
          if (!visited[nextEdge.v2 - 1]) {
            queue.offer(nextEdge);
          }
        }
      }
    }

    if (mst.size() != v - 1) {
      System.out.println("Graph is not connected. MST cannot be formed.");
    } else {
      System.out.println("MST: ");
      for (GraphEdge edge : mst) {
        System.out.println(edge.v1 + " " + edge.v2 + " " + edge.w);
      }
    }
  }
}

class GraphEdge implements Comparable<GraphEdge> {
  int v1;
  int v2;
  int w;

  public GraphEdge(int v1, int v2, int w) {
    this.v1 = v1;
    this.v2 = v2;
    this.w = w;
  }

  public int compareTo(GraphEdge that) {
    return this.w - that.w;
  }
}
